### ISSUE TITLE
Unified API for getting and setting variables.

### ISSUE THREAD

**Issue by jackjansen** [cwi-dis/iotsa#14]
The API for BLE servers looks somewhat like the API for config variables. The API for REST/COAP interfaces is completely different, but has a very similar functionality. And there's basically no API for creating a user-readable web interface.

These need to be unified (at least optionally).

**Comment by jackjansen** [comment]
A first step would be to have a similar API for config file, Web request and REST request. Basically copy the `get()` and `put()` methods from configFile, with optional checking of having the right mode/credentials.

**Comment by jackjansen** [comment]
Actually #35 is also part of the solution. This allows treating application-defined "objects" within a module in a unified way.

**Comment by jackjansen** [comment]
What I want to get at is having a fully rest/coap/web interface/ble/ota enabled _Blink_ example something like

```
#include "iotsa.h"

#define PIN_LED 13

IotsaApplication app("Blink");
IotsaMainModule mod(app, "Blink settings");

int duration;

void setup() {
    app.setup();
    auto _duration = mod.add_variable(duration, "duration");
    _duration.set_description("Blink on/off duration (ms)");
    pinMode(PIN_LED, OUTPUT);
}

void loop() {
  app.loop();
  digitalWrite(PIN_LED, HIGH);
  delay(duration);
  digitalWrite(PIN_LED, LOW);
  delay(duration);
}
```

(Ignoring for the moment that that would wreak havoc with WiFi, because of the use of `delay()`, but this should be the general structure).

This would create the user interface web page with a form, the REST/COAP get/put, configfile save/restore, etc.
Probably `add_variable` should return something through which you could set options (like whether or not to save to the config file, or whether something is a password, or its BLE UUID and all that).

**Comment by jackjansen** [comment]
Sequencing note: #40 (investigate always-on `IOTSA_WITH_PLACEHOLDERS`) is a soft prerequisite — #40's own text notes it would make a unified base module, and so this issue, more feasible. Worth landing #40 first.

— Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Sequencing note (2026-08-25), supersedes the earlier "#40 is a soft prerequisite" note:
this issue now depends on the #204 → #205 → #206 chain instead. #204 defines what the
unified get/put surface actually covers (module+feature tiers); #205 cleans up the
`IOTSA_WITH_*`/`IOTSA_WITHOUT_*` flags gating it; #206 provides the class mechanism
(folding `IotsaApiProvider`/`IotsaBLEApiProvider` into `IotsaMod`) a unified base module
would actually build on. #40 is closed but turned out to be only tangentially related.
Full reasoning on cwi-dis/iotsa#199.

— Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Attribute vocabulary (2026-08-25), worked out while discussing #204 — kept here rather
than there, since #204 is scoped to module/feature *tiers* (the lifecycle axis) and this
is a different, orthogonal axis: what a module actually exposes.

**Module vs. attribute are two separate concepts, not a hierarchy.** A module
(`IotsaBaseMod`) is a lifecycle citizen: `setup()`/`loop()`/`serverSetup()`, registered
with `IotsaApplication`. An **attribute** is a named piece of *state* something exposes
through the uniform get/set/persist/render surface (REST/CoAP/BLE/web-form/config-file)
— owned by a module, or nested inside another attribute. Two kinds:

- **Simple attribute** — a scalar (int/bool/String/float/enum). This is exactly what the
  Blink sketch above (`mod.add_variable(duration, "duration")`) is reaching for: declare
  it once, get the REST/CoAP/BLE/web-form/config surface for free.
- **Structured attribute** — a composite of other attributes (simple or structured).
  `IotsaModObject`/`IotsaApiModObject` (`src/iotsa.h`, `src/iotsaApi.h`) is exactly this
  shape today: `configLoad`/`configSave` (persistence), `getHandler`/`putHandler`
  (REST/CoAP/BLE), `formHandler_*` (web-form) — deliberately no `setup()`/`loop()`/
  `serverSetup()`, because a structured attribute has no lifecycle of its own, unlike a
  module. Five real adopters confirmed by grep across the whole iotsa-group family:
  `IotsaUser` (`src/iotsaMultiUser.h`), `IotsaRequest` (`src/iotsaRequest.h`, already
  flagged elsewhere in memory as similar in shape to `IotsaBLEClientMod` — worth
  revisiting once this lands), `IotsaStaticTokenObject` (`examples/HelloToken`),
  `Iotsa433Telegram` (`iotsa433`), and lissabon's `AbstractDimmer`. In every case today
  the composite's get/put/persist/render methods are hand-written by hand per class,
  rather than composed automatically from the attribute's constituent parts — that's the
  gap a real "structured attribute" API would close.

**A collection is just an array-valued structured attribute, and that shape is
currently hand-rolled twice.** `IotsaMultiUserMod` owns `std::vector<IotsaUser> users`
and hand-writes its own get/put/config/web fan-out across the vector. Lissabon's
`Lissabon::DimmerCollection` (`libLissabon/src/DimmerCollection.h`) is a
`std::vector<AbstractDimmer*>` doing the exact same job over a different item type,
built completely independently. Generalizing "array of structured attribute" once would
retire both and be immediately reusable for any future "module manages a list of
like-shaped things" (BLE known-devices, capabilities, files, ...).

— Claude Code (Claude Sonnet 5)

**Comment by jackjansen** [comment]
Deferred to v4.0. Sized up the design + implementation work needed (simple/structured/collection attribute API, REST+CoAP vs. BLE reconciliation, the #107 auth hook, wire-compat with deployed devices, migrating real modules to prove the shape) against #106 as a yardstick -- it's at least that scale, likely bigger, and it carries an open dependency on undesigned #107 plus an open interaction with undesigned lissabon#25. Consistent with the 2026-09-01 decision that already pushed layers 2/3 out of the v3.0 gate. lissabon#2 (which depends on this issue's shape) deferred alongside it.

— Claude Code (Claude Sonnet 5)
